File: search/binary_search.go
package search

const (
	NotFound = -1
)

func BinarySearch(A []int, x int) int {
	var (
		N    = len(A)
		low  = 0
		high = N - 1
		mid  int
	)
	if N > 0 && (x < A[low] || x > A[high]) {
		return NotFound
	}
	for low <= high {
		mid = (low + high) / 2
		m := A[mid]
		if x > m {
			low = mid + 1
		} else if x < m {
			high = mid - 1
		} else {
			// found
			return mid
		}
	}
	return NotFound
}

File: book_data_structure_and_algorithm_analysis/ch01/binary_search.c
#include "binary_search.h"

int BinarySearch(const ElementType A[], ElementType x, int N) {
    int Low = 0, Mid, High = N - 1;
    if (N > 0 && (x < A[Low] || x > A[High])) {
        return NotFound;
    }
    while (Low <= High) {
        Mid = (Low + High) / 2;
        int y = A[Mid];
        if (y < x) {
            Low = Mid + 1;
        } else if (y > x) {
            High = Mid - 1;
        } else {
            /* Found */
            return Mid;
        }
    }
    return NotFound;
}
